source/distributed/tria.cc

```diff
@@ -1326,6 +941,173 @@ namespace
     // get the weight, increment the pointer, and return the weight
     return *this_object->current_pointer++;
   }
+
+
+
+  // Note: We will no longer need the quadrant pointer in this alias
+  //       if no functionality accesses quadrants' user_data anymore, i.e.
+  //       if we do not support p4est lower than 2.0.
+  template <int dim, int spacedim>
+  using quadrant_cell_relation_t = typename std::tuple<
+    typename dealii::internal::p4est::types<dim>::quadrant *,
+    typename dealii::Triangulation<dim, spacedim>::CellStatus,
+    typename dealii::Triangulation<dim, spacedim>::cell_iterator>;
+
+
+
+  template <int dim, int spacedim>
+  inline void
+  add_single_quadrant_cell_relation(
+    std::vector<quadrant_cell_relation_t<dim, spacedim>> &      quad_cell_rel,
+    const typename dealii::internal::p4est::types<dim>::tree &  tree,
+    const unsigned int                                          idx,
+    const typename Triangulation<dim, spacedim>::cell_iterator &dealii_cell,
+    const typename Triangulation<dim, spacedim>::CellStatus     status)
+  {
+    const unsigned int local_quadrant_index = tree.quadrants_offset + idx;
+
+    // Note: We will no longer need the quadrant pointer
+    //       if no functionality accesses quadrants' user_data anymore, i.e.
+    //       if we do not support p4est lower than 2.0.
+    const auto q =
+      static_cast<typename dealii::internal::p4est::types<dim>::quadrant *>(
+        sc_array_index(const_cast<sc_array_t *>(&tree.quadrants), idx));
+
+    // check if we will be writing into valid memory
+    Assert(local_quadrant_index < quad_cell_rel.size(), ExcInternalError());
+
+    // store relation
+    quad_cell_rel[local_quadrant_index] =
+      std::make_tuple(q, status, dealii_cell);
+  }
+
+
+
+  template <int dim, int spacedim>
+  void
+  compute_quadrant_cell_relations_recursively(
+    std::vector<quadrant_cell_relation_t<dim, spacedim>> &        quad_cell_rel,
+    const typename dealii::internal::p4est::types<dim>::tree &    tree,
+    const typename Triangulation<dim, spacedim>::cell_iterator &  dealii_cell,
+    const typename dealii::internal::p4est::types<dim>::quadrant &p4est_cell)
+  {
+    // find index of p4est_cell in the quadrants array of the corresponding tree
+    const int idx = sc_array_bsearch(
+      const_cast<sc_array_t *>(&tree.quadrants),
+      &p4est_cell,
+      dealii::internal::p4est::functions<dim>::quadrant_compare);
```

Comment:
I was originally concerned that doing a bsearch on all cells is going to be too expensive. But I realize now that the bsearch was also in the original code where we had never noticed that it is slow. So this must be ok.